internal/wasm/function_definition.go

```diff
@@ -102,13 +102,31 @@ func (m *Module) BuildFunctionDefinitions() {
 			}
 		}
 	}
+
+	if len(m.FunctionDefinitionSection) == 0 || m.DWARFLines == nil {
+		return
+	}
+	switch m.DWARFLines.Language() {
+	case wasmdebug.DwarfLangRust:
+		m.DWARFLines.LinkageNamesDo(func(linkageName, name string) {
+			for i := 0; i < len(m.FunctionDefinitionSection); i++ {
```

Comment:
There may be opportunities for improvement by keeping the index from one call to another, which would lower the number of iterations if the DIEs are roughly in the same order as how functions are laid out in the WASM module.

Reply:
I wonder how slow could this be in reality - the sqlite3 binary has roughly 1.5k functions, and I guess this could be really slow https://github.com/ncruces/go-sqlite3/blob/main/embed/sqlite3.wasm

Reply:
we have a benchmark right? just @pelletier needs to manually download the sqlite wasm iirc to replace internal/integration_test/bench/testdata/case.wasm then run Benchmark in internal/integration_test/bench/bench_test.go right?

Reply:
we have a bench but dwarf is not available in these checked-in binaries as it is huge.

Reply:
https://github.com/ncruces/go-sqlite3/blob/main/embed/sqlite3.wasm has dwarf tho right? for running bench offline I mean (I didn't mean to commit the sqlite to this PR, rather run it to check)

Reply:
plus the case.wasm is in reality not that big compared to sqlite3 - the number of functions 10x smaller and this for-loop can be O(N^2) in worst case maybe

Reply:
> which would lower the number of iterations if the DIEs are roughly in the same order as how functions are laid out in the WASM module.

on this point, I remember Zig produces the "weird" dwarf as I added in the comment in this file. Maybe this assumption could be wrong

```
 		// Note that the dwarf.LineEntry.SeekPC API shouldn't be used because the Go's dwarf package assumes 
		// all the line entries in an Entry are sorted in increasing order which *might not* be true
		// for some languages. Such order requirement is not a part of DWARF specification,
		// and in fact Zig language tends to emit interleaved line information.
```

Reply:
for reducing likelihood of worst case, we can do the normal map lookup thing right?  We can also move the map somewhere it is re-used. We have lookups elsewhere during compilation though agree maybe this would be a larger one.

style: I personally would not use a callback rather imperative and a map to prevent worst case (knowing it will allocate to the size of max functions). I think we rarely, if ever, make do/callback style in our code regardless of locking. It would be cheaper to stay imperative as well, so that helps with the cost.